### PR TITLE
Task2 router api

### PR DESCRIPTION
**Pull request by DariaRusovich**

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,12 +1,43 @@
...
+            </Route>
+            <PrivateRoute path="/department/:id/employees">
+              <EmployeesPage></EmployeesPage>
+            </PrivateRoute>
+            <PrivateRoute exact path="/">
+              <HomePage></HomePage>
+            </PrivateRoute>
+            <Route path="*">
+              <NotFoundPage />
+            </Route>
```
Comment (vitalii235) [inline src/App.js]: make realisation by using map
```suggestion
            privateRoutes.map(route => <PrivateRoute path={"route.path"} key={"..."}>{route.children}</PrivateRoute>)
```
the same for typical routes, it allows you make changes without code duplication

File: src/App.js [hunk 2]
```diff
@@ -1,12 +1,43 @@
...
-
-
+import React, { Component } from 'react';
+import { Route, Switch, Link } from 'react-router-dom';
+import Header from './components/Header';
+import LoginPage from './pages/LoginPage';
+import HomePage from './pages/HomePage';
+import EmployeesPage from './pages/EmployeesPage';
+import NotFoundPage from './pages/NotFoundPage';
+import PrivateRoute from './components/PrivateRoute';
```
Comment (vitalii235) [inline src/App.js]: PrivateRoute should not be in components folder, create another folder for different route types

File: src/api/api.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+export async function getDepartmentById(id) {
+  return api.get(`/departments/${id}`);
+}
+
+// export async function getEmployeeByDepartmentId(id) {
+//   return api.get(`/departments/${id}/employees`);
+// }
+export async function signin(loginData) {
+  return api.post('/login', loginData);
+}
```
Comment (vitalii235) [inline src/api/api.js]: Its okay to use functions but more common practice if you use class instead and create all methods inside

File: src/api/api.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import axios from 'axios';
+
+const api = axios.create({
+  baseURL: 'http://localhost:3000/api/v1',
```
Comment (vitalii235) [inline src/api/api.js]: better create constants common file, where you save all strings what may be changed or reused and export it from there

```suggestion
  baseURL:  BASE_URL,
```

Comment (vitalii235) [inline src/api/api.js]: And BASE_URL gonna be taken from local environments

File: src/components/Department.jsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import { withRouter } from 'react-router-dom';
+import { dateFormat } from '../utils/dateFormatter';
+
+
+
+
+class Department extends Component {
+  constructor(props) {
+    super(props);
+  }
```
Comment (vitalii235) [inline src/components/Department.jsx]: You may not use constructor, this is an old syntax

File: src/components/Department.jsx [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  render() {
+    const { department } = this.props;
+    return (
+      <>
+
```
Comment (vitalii235) [inline src/components/Department.jsx]: Remove unused spaces

File: src/components/Department.jsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            <p className="department-item__description">
+              {department.description}
+            </p>
+            <span className="department-item__date">
+              {dateFormat(department.created_at)}
+            </span>
+            {/* <span className='department-item__date'>{department.updated_at}</span> */}
+            <button className="btn">Edit</button>
+            <button className="btn">Delete</button>
+            <Link to={`/department/${department.id}/employees`}>
```
Comment (vitalii235) [inline src/components/Department.jsx]: Also, i would rather use constant here, imagine you use this string in many places, but tomorrow the url is changed, if you use constants you have to change this just in one place

File: src/components/DepartmentsList.jsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+export default class DepartmentsList extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      departments: [],
+      isLoading: false,
+      error: null,
+    };
+  }
```
Comment (vitalii235) [inline src/components/DepartmentsList.jsx]: You also may not use the constructor here and use state as property
```suggestion
     state = {...}
```

File: src/components/Department.jsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+              <button className="btn">Employees</button>
```
Comment (vitalii235) [inline src/components/Department.jsx]: Why do you need this button if it does not make any click events or actions?

File: src/components/DepartmentsList.jsx [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+  async componentDidMount() {
+    this.setState({ isLoading: true });
+    setTimeout(async () => {
```
Comment (vitalii235) [inline src/components/DepartmentsList.jsx]: Why do you use setTimeout here? Do you have some specific task for this?

File: src/components/DepartmentsList.jsx [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+      this.setState({ isLoading: false });
+    }, 500);
+  }
+
+  render() {
+    const { departments, isLoading, error } = this.state;
+    //console.log(departments);
+
+    if (isLoading) {
+      return <Loader></Loader>;
```
Comment (vitalii235) [inline src/components/DepartmentsList.jsx]:
```suggestion
      return <Loader/>;
```

File: src/components/Employee.jsx [hunk 12]
```diff
@@ -0,0 +1,35 @@
+import React, { Component } from 'react';
+import { dateFormat } from '../utils/dateFormatter';
+
+export default class Employee extends Component {
+  constructor(props) {
+    super(props);
+  }
```
Comment (vitalii235) [inline src/components/Employee.jsx]: the same here with constructor, its no needed

File: src/components/Employee.jsx [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+  render() {
+    const { employee } = this.props;
+    return (
+      <>
+        {employee && (
```
Comment (vitalii235) [inline src/components/Employee.jsx]: I would not set this condition right here. Why component what does not has any logic should know about any condition, it should be responsive just for rendering some content
```suggestion
{employee && <Employe employee={employee}/>}
```
I would do this before the call

File: src/components/Employee.jsx [hunk 14]
```diff
@@ -0,0 +1,35 @@
+import React, { Component } from 'react';
+import { dateFormat } from '../utils/dateFormatter';
+
+export default class Employee extends Component {
+  constructor(props) {
+    super(props);
+  }
+  render() {
+    const { employee } = this.props;
+    return (
```
Comment (vitalii235) [inline src/components/Employee.jsx:8]: Here you may do employee destructuring with default values and not use below employee.username, employee.first_name ...

File: src/components/EmployeesList.jsx [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+import { withRouter } from 'react-router-dom';
+import { getDepartmentById, getEmployeeByDepartmentId } from '../api/api';
+
+import Employee from './Employee';
+import Loader from './Loader';
+
+class EmployeesList extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
```
Comment (vitalii235) [inline src/components/EmployeesList.jsx]: The same here, constructor is no needed, and state may be called as just property

File: src/components/EmployeesList.jsx [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+      departments: {},
+      employees: [],
+      isLoading: false,
+      error: null,
+    };
+  }
+
+  // async componentDidMount() {
```
Comment (vitalii235) [inline src/components/EmployeesList.jsx]: Remove unused code

File: src/components/EmployeesList.jsx [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+  //     } else {
+  //       this.setState({ error: employeesDataError });
+  //     }
+  //     this.setState({ isLoading: false });
+  //   }, 500);
+  // }
+
+  async componentDidMount() {
+    this.setState({ isLoading: true });
+    setTimeout(async () => {
```
Comment (vitalii235) [inline src/components/EmployeesList.jsx]: What is the specific task for setTimeout here?

File: src/components/EmployeesList.jsx [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+    if (!employees) {
+      return (
+        <h2>
+          No employees in the {department} department.
+          <Link to="/"> Go back.</Link>
+        </h2>
+      );
+    }
+    if (isLoading) {
+      return <Loader></Loader>;
```
Comment (vitalii235) [inline src/components/EmployeesList.jsx]:
```suggestion
      return <Loader/>;
```

File: src/components/EmployeesList.jsx [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+          <div className="container section-wrap">
+            {employees && (
+              <h2>
+                {employees.length} employees in the {department} department
+              </h2>
+            )}
+            <div className="employees-list">
+              {employees &&
+                employees.map((employee) => (
+                  <Employee key={employee.id} employee={employee}></Employee>
```
Comment (vitalii235) [inline src/components/EmployeesList.jsx]:
```suggestion
                  <Employee key={employee.id} employee={employee}/>
```

File: src/components/LoginForm.jsx [hunk 20]
```diff
@@ -0,0 +1,27 @@
+import React, { Component } from 'react';
+
+export default class LoginForm extends Component {
+  constructor(props) {
+    super(props);
```
Comment (vitalii235) [inline src/components/LoginForm.jsx]: The same here, constructor is no needed

File: src/components/LoginForm.jsx [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+    this.handleSubmit = this.handleSubmit.bind(this)
```
Comment (vitalii235) [inline src/components/LoginForm.jsx]: If you use arrowFunction you dont need use bind

File: src/components/LoginForm.jsx [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+    handleSubmit(event){
```
Comment (vitalii235) [inline src/components/LoginForm.jsx]: About comment above
```suggestion
    handleSubmit = (event) =>{
```

File: src/components/LoginForm.jsx [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+        event.preventDefault()
+        const loginData = {}
+        loginData.username = event.target.username.value
+        loginData.password = event.target.password.value
```
Comment (vitalii235) [inline src/components/LoginForm.jsx]: In the future you probably need to add some validation, so you have to do those inputs are controlled, if not its okay

File: src/components/PrivateRoute.jsx [hunk 24]
```diff
@@ -0,0 +1,28 @@
+import React, { Component } from 'react';
+import { Route, Redirect } from 'react-router-dom';
+
+export default class PrivateRoute extends Component {
+    constructor(props){
```
Comment (vitalii235) [inline src/components/PrivateRoute.jsx]: The same here

File: src/pages/LoginPage.jsx [hunk 25]
```diff
@@ -0,0 +1,36 @@
+import React, { Component } from 'react';
+import { signin } from '../api/api';
+import { Redirect } from 'react-router-dom';
+import LoginForm from '../components/LoginForm';
+
+
+export default class LoginPage extends Component {
+  constructor(props) {
```
Comment (vitalii235) [inline src/pages/LoginPage.jsx]: The same with constructor and state here

File: src/pages/LoginPage.jsx [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+    super(props);
+    this.state = {
+      token: null,
+    };
+    this.signIn = this.signIn.bind(this)
+  }
+
+
+  async signIn(loginData) {
```
Comment (vitalii235) [inline src/pages/LoginPage.jsx]: If you use arrow function, you dont need bind this function above

```suggestion
  async signIn = (loginData) => {
```

File: src/pages/LoginPage.jsx [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+    setTimeout(async () => {
```
Comment (vitalii235) [inline src/pages/LoginPage.jsx]: Why do you need the timeout here? Some specific task?

File: src/components/DepartmentsList.jsx (hunk 11 above)
Comment (vitalii235) [inline src/components/DepartmentsList.jsx]: I would add Loader for global layout, and to some global state, so you are able from any function make toggle, and no worries where its gonna be rendered

File: src/components/DepartmentsList.jsx (hunk 10 above)
Comment (DariaRusovich) [reply src/components/DepartmentsList.jsx]: I used the setTimeout to show how the spinner would work

File: src/components/EmployeesList.jsx (hunk 17 above)
Comment (DariaRusovich) [reply src/components/EmployeesList.jsx]: I use the setTimeout to show how spinner would work

File: src/components/LoginForm.jsx (hunk 22 above)
Comment (DariaRusovich) [reply src/components/LoginForm.jsx]: Cool)) o_O

File: src/components/DepartmentsList.jsx (hunk 10 above)
Comment (vitalii235) [reply src/components/DepartmentsList.jsx]: Got it, but i think it may be removed)))

File: src/components/EmployeesList.jsx (hunk 17 above)
Comment (vitalii235) [reply src/components/EmployeesList.jsx]: Got it, but i think it may be removed)))

File: src/components/Department.jsx (hunk 9 above)
Comment (DariaRusovich) [reply src/components/Department.jsx]: The button seems to be by design

File: src/components/Employee.jsx (hunk 14 above)
Comment (DariaRusovich) [reply src/components/Employee.jsx:8]: I don't get it.

File: src/pages/LoginPage.jsx (hunk 26 above)
Comment (DariaRusovich) [reply src/pages/LoginPage.jsx]: The asynchronous method does not work

File: src/components/Employee.jsx [hunk 28]
```diff
@@ -0,0 +1,35 @@
+import React, { Component } from 'react';
+import { dateFormat } from '../utils/dateFormatter';
+
+export default class Employee extends Component {
+  constructor(props) {
+    super(props);
+  }
+  render() {
+    const { employee } = this.props;
```
Comment (vitalii235) [inline src/components/Employee.jsx:9]:
```suggestion
    const { employee = {} } = this.props;
    const {username, first_name, last_name, email, created_at} = employee
```

File: src/pages/LoginPage.jsx (hunk 26 above)
Comment (vitalii235) [reply src/pages/LoginPage.jsx]: Sorry, my fault signIn = async (loginData) => {

File: src/components/Employee.jsx (hunk 14 above)
Comment (vitalii235) [reply src/components/Employee.jsx:8]: const { employee = {} } = this.props;
    const {username, first_name, last_name, email, created_at} = employee

File: src/App.js [hunk 29]
```diff
@@ -1,12 +1,26 @@
...
+import { Switch, Link } from 'react-router-dom';
+import Header from './components/Header';
+import { privateRouteComponents, routeComponents } from './constants/routes';
 
 export default class App extends Component {
   render() {
     return (
       <>
-     
+        <Header></Header>
```
Comment (vitalii235) [inline src/App.js]:
```suggestion
        <Header/>
```

File: src/pages/LoginPage.jsx [hunk 30]
```diff
@@ -0,0 +1,27 @@
...
+  state = {
+    token: null,
+  };
+
+  signIn = async (loginData) => {
+    const [userDataError, userData] = await signin(loginData);
+    const token = userData.USERTOKEN || userData.ADMINTOKEN;
+    if (!userDataError && userData.USERTOKEN) {
+      localStorage.setItem('token', token);
+      this.setState({ token });
```
Comment (vitalii235) [inline src/pages/LoginPage.jsx]: why do you need to set token to state?

File: src/protectedRoutes/PrivateRoute.jsx [hunk 31]
```diff
@@ -0,0 +1,27 @@
+import React, { Component } from 'react';
+import { Route, Redirect } from 'react-router-dom';
+
+export default class PrivateRoute extends Component {
+  render() {
+    const { children, ...rest } = this.props;
+
+    console.log('rest', rest);
```
Comment (vitalii235) [inline src/protectedRoutes/PrivateRoute.jsx]: remove console plz)

File: src/components/EmployeesList.jsx [hunk 32]
```diff
@@ -0,0 +1,73 @@
...
+        employees: department.departmentById.employees,
+      });
+    } else {
+      this.setState({ error: departmentError });
+    }
+    this.setState({ isLoading: false });
+  }
+  render() {
+    const { department, employees, isLoading, error } = this.state;
+    console.log(department, employees);
```
Comment (vitalii235) [inline src/components/EmployeesList.jsx]: remove console plz)

File: src/pages/LoginPage.jsx (hunk 30 above)
Comment (DariaRusovich) [reply src/pages/LoginPage.jsx]: I get the token in this component and set it to the state, and initially the token is null, below is a check if the token  received redirect to '/'

File: src/pages/LoginPage.jsx [hunk 33]
```diff
@@ -0,0 +1,23 @@
...
+
+class LoginPage extends Component {
+  signIn = async (loginData) => {
+    const [userDataError, userData] = await signin(loginData);
+    const token = userData.USERTOKEN || userData.ADMINTOKEN;
+    if (!userDataError && userData.USERTOKEN) {
+      localStorage.setItem('token', token);
+      this.props.history.push('/');
+    } else {
+      this.setError(userDataError);
```
Comment (vitalii235) [inline src/pages/LoginPage.jsx]:
```suggestion
      this.props.setError(userDataError);
```

File: src/components/LoginForm.jsx [hunk 34]
```diff
@@ -0,0 +1,33 @@
+import React, { Component } from 'react';
+
+export default class LoginForm extends Component {
+  handleSubmit = (event) => {
+    event.preventDefault();
+    const loginData = {};
+    loginData.username = event.target.username.value.trim();
+    loginData.password = event.target.password.value.trim();
+    const validUsername = loginData.username.length > 3;
+    const validpassword = loginData.password.length > 3;
```
Comment (vitalii235) [inline src/components/LoginForm.jsx]: Create separate service for validation,  this conditions shoud not be set in function...

File: src/api/api.js [hunk 35]
```diff
@@ -0,0 +1,43 @@
...
+
+export async function getDepartments() {
+  return api.get(DEPARTMENTS_URL);
+}
+export async function getDepartmentById(id) {
+  return api.get(`${DEPARTMENT_BY_ID_URL}${id}`);
+}
+export async function signin(loginData) {
+  return api.post(LOGIN, loginData);
+}
```
Comment (vitalii235) [inline src/api/api.js]: Please create this service by using class, not just bunch of functions

**Review by vitalii235 — APPROVED** [review]

**Pull request merged**